File: app/src/main/java/com/ponthaitay/lifecycleactivity/googleio2017/lifecycle/javacode/MainActivity.java
package com.ponthaitay.lifecycleactivity.googleio2017.lifecycle.javacode;

import android.Manifest;
import android.annotation.SuppressLint;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.util.Log;
import android.widget.TextView;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionDeniedResponse;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;
import com.ponthaitay.lifecycleactivity.googleio2017.R;

import java.util.List;

public class MainActivity extends BaseLifecycleActivity {

    private static final String TAG = MainActivity.class.getName();
    private MyLocationManager locationManager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        locationManager = new MyLocationManager(this, getLifecycle(), new MyLocationManager.MyLocationListener() {

            @SuppressLint("DefaultLocale")
            @Override
            public void onLocationChange(Location location) {
                Log.d(TAG, location.toString());
                String result = String.format("%f / %f", location.getLatitude(), location.getLongitude());
                ((TextView) findViewById(R.id.tv_result)).setText(result);
            }

            @Override
            public void onProviderDisabled(String msg) {
                Log.d(TAG, msg);
            }

            @Override
            public void onProviderEnabled(String provider) {
                Log.d(TAG, provider);
            }
        });

        checkPermissionLocation();
    }

    private void checkPermissionLocation() {
        Dexter.withActivity(this)
                .withPermissions(Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION)
                .withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {
                        if (!hasDeniedPermission(report)) {
                            locationManager.locationEnable();
                        } else {
                            Snackbar.make(findViewById(R.id.container), report.toString(), Snackbar.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(List<PermissionRequest> list, PermissionToken permissionToken) {

                    }

                    private boolean hasDeniedPermission(MultiplePermissionsReport report) {
                        List<PermissionDeniedResponse> denyPermission = report.getDeniedPermissionResponses();
                        return denyPermission != null && denyPermission.size() > 0;
                    }
                });
    }
}

File: app/src/main/java/com/ponthaitay/lifecycleactivity/googleio2017/lifecycle/kotlincode/MainActivity.kt
package com.ponthaitay.lifecycleactivity.googleio2017.lifecycle.kotlincode

import android.Manifest
import android.annotation.SuppressLint
import android.location.Location
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.util.Log
import android.widget.TextView
import com.karumi.dexter.Dexter
import com.karumi.dexter.MultiplePermissionsReport
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.PermissionRequest
import com.karumi.dexter.listener.multi.MultiplePermissionsListener
import com.ponthaitay.lifecycleactivity.googleio2017.R


class MainActivity : BaseLifecycleActivity() {

    var locationManager: MyLocationManager? = null
    var container: android.view.View? = null
    var TAG: String = javaClass.simpleName
    var tvResult: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main_lifecycle)
        container = findViewById(R.id.container)
        tvResult = findViewById(R.id.tv_result) as TextView?
        locationManager = MyLocationManager(this, lifecycle, object : MyLocationManager.MyLocationListener {
            override fun onProviderDisabled(msg: String?) {
                Log.d(TAG, "{$msg}")
            }

            override fun onProviderEnabled(provider: String?) {
                Log.d(TAG, "{$provider}")
            }

            @SuppressLint("SetTextI18n")
            override fun onLocationChange(location: Location?) {
                Log.d(TAG, "{$location}")
                tvResult!!.text = "${location?.latitude}" + "/" + "${location?.longitude}"
            }
        })

        checkPermissionLocation()
    }

    private fun checkPermissionLocation() {
        Dexter.withActivity(this)
                .withPermissions(Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION)
                .withListener(object : MultiplePermissionsListener {
                    override fun onPermissionsChecked(report: MultiplePermissionsReport) {
                        if (!hasDeniedPermission(report)) {
                            locationManager!!.locationEnable()
                        } else {
                           Snackbar.make(container!!, report.toString(), Snackbar.LENGTH_SHORT).show()
                        }
                    }

                    override fun onPermissionRationaleShouldBeShown(permissions: MutableList<PermissionRequest>, token: PermissionToken) {
                        token.continuePermissionRequest()
                    }

                    private fun hasDeniedPermission(report: MultiplePermissionsReport?): Boolean {
                        return report?.deniedPermissionResponses?.size!! > 0
                    }

                }).check()
    }
}